QgsFeature python optimisations

nyalldawson opened the pull request:
Optimise some more frequently called methods for use in PyQGIS

nyalldawson committed:
Optimise QgsAttributes python constructor

Avoid multiple checks for detachment for each attribute

nyalldawson committed:
Add more HoldGIL annotations to cheap methods

troopa81 commented on python/PyQt6/core/auto_generated/qgsattributes.sip.in:
```diff
@@ -72,14 +72,15 @@ typedef QVector<QVariant> QgsAttributes;
     if ( obj == Py_None )
     {
-      qv->append( QVariant( QVariant::Int ) );
+      *outData++ = QVariant( QVariant::Int );
```
Are we gaining something relevant here ?

nyalldawson replied:
Yes -- it reduces the qt implicitly shared container detachment checks from happening once per attribute appended to just once upfront. Ie both QVector::append and QVector::data check for shared use and detachment **every** time they are called, even if it's in a trivial loop like this. These detachment checks add up quickly!

ptitjano commented:
@nyalldawson Thanks. Have you been able to test it and measure the improvement?

nyalldawson committed:
Additional shortcut to QgsAttributes creation from Python

Avoid constructing a bunch of default QVariant objects which
we will just replace immediately

nyalldawson commented:
@ptitjano 

> Have you been able to test it and measure the improvement?

With a similar test to the loop from https://github.com/qgis/QGIS/pull/57470, calling 10.5 million times

Before: 
18.3s

After:
16.4s

With the optimisations from https://github.com/qgis/QGIS/pull/57470:

Before (#57470 alone):
12.8s

After (this + #57470):
11.7s


So no where near as dramatic as #57470, but still worthwhile!

troopa81 commented on python/PyQt6/core/auto_generated/qgsattributes.sip.in:
```diff
@@ -70,16 +70,21 @@ typedef QVector<QVariant> QgsAttributes;
+  // Initialize attributes to null. This has two motivations:
+  // 1. It speeds up the QVector construction, as otherwise we are creating n default QVariant objects (default QVariant constructor is not free!)
+  // 2. It lets us shortcut in the loop below when a Py_None is encountered in the list
+  const QVariant nullVariant = QVariant( QVariant::Int );
```
```suggestion
  const QVariant nullVariant( QVariant::Int );
```

nyalldawson committed:
Fix bad merge

nyalldawson commented:
@troopa81 good to merge now?
